azure-iot-device/tests/common/pipeline/test_pipeline_stages_mqtt.py

```diff
@@ -107,28 +107,91 @@ def transport(mocker):
 
 
 @pytest.fixture
-def op_set_connection_args(callback):
+def op_set_connection_args(mocker):
     return pipeline_ops_mqtt.SetMQTTConnectionArgsOperation(
         client_id=fake_client_id,
         hostname=fake_hostname,
         username=fake_username,
         ca_cert=fake_ca_cert,
         client_cert=fake_certificate,
-        callback=callback,
+        sas_token=fake_sas_token,
+        callback=mocker.MagicMock(),
     )
 
 
+@pytest.fixture
+def op_connect(mocker):
+    return pipeline_ops_base.ConnectOperation(callback=mocker.MagicMock())
+
+
+@pytest.fixture
+def op_reconnect(mocker):
+    return pipeline_ops_base.ReconnectOperation(callback=mocker.MagicMock())
+
+
+@pytest.fixture
+def op_disconnect(mocker):
+    return pipeline_ops_base.DisconnectOperation(callback=mocker.MagicMock())
+
+
+@pytest.fixture
+def op_publish(mocker):
+    return pipeline_ops_mqtt.MQTTPublishOperation(
+        topic=fake_topic, payload=fake_payload, callback=mocker.MagicMock()
+    )
+
+
+@pytest.fixture
+def op_subscribe(mocker):
+    return pipeline_ops_mqtt.MQTTSubscribeOperation(topic=fake_topic, callback=mocker.MagicMock())
+
+
+@pytest.fixture
+def op_unsubscribe(mocker):
+    return pipeline_ops_mqtt.MQTTUnsubscribeOperation(topic=fake_topic, callback=mocker.MagicMock())
+
+
+@pytest.fixture
+def create_transport(stage, transport, op_set_connection_args):
+    stage.run_op(op_set_connection_args)
+
+
+# TODO: This should be a package level class inherited by all .run_op() tests in all stages
+class RunOpTests(object):
+    @pytest.mark.it(
+        "Completes the operation with failure if an unexpected Exception is raised while executing the operation"
+    )
+    def test_completes_operation_with_error(self, mocker, stage):
+        execution_exception = Exception()
+        mock_op = mocker.MagicMock()
+        stage._execute_op = mocker.MagicMock(side_effect=execution_exception)
+
+        stage.run_op(mock_op)
+        assert mock_op.error is execution_exception
+
+    @pytest.mark.it(
+        "Allows any BaseException that was raised during execution of the operation to propogate"
+    )
+    def test_base_exception_propogates(self, mocker, stage):
+        execution_exception = BaseException()
+        mock_op = mocker.MagicMock()
+        stage._execute_op = mocker.MagicMock(side_effect=execution_exception)
+
+        with pytest.raises(BaseException):
+            stage.run_op(mock_op)
+
+
 @pytest.mark.describe(
     "MQTTTransportStage - .run_op() -- called with pipeline_ops_mqtt.SetMQTTConnectionArgsOperation"
 )
-class TestMQTTProviderRunOpWithSetConnectionArgs(object):
+class TestMQTTProviderRunOpWithSetConnectionArgs(RunOpTests):
     @pytest.mark.it("Creates an MQTTTransport object")
     def test_creates_transport(self, stage, transport, op_set_connection_args):
         stage.run_op(op_set_connection_args)
```

Comment:
>run_op [](start = 14, length = 6)

Turns out I was able to test only on the wrapping `.run_op()` method rather than the inner `._execute_op()` after all. Just took the right approach to writing the tests.